---
title: "Metagenomics Summary Report"
# author: "Stephen Turner"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    code_download: true
    code_folding: hide
    # out.width: "100%"
    fig.width: 9
    fig.height: 3
    warning: FALSE
    message: FALSE
    # theme: default
    # theme: cerulean
    # theme: journal
    # theme: flatly
    # theme: readable
    # theme: spacelab
    # theme: united
    # theme: cosmo
    # theme: lumen
    # theme: paper
    # theme: sandstone
    # theme: simplex
    # theme: yeti
    toc: yes
    toc_depth: 1
    toc_float:
      collapsed: false
      smooth_scroll: false
subtitle: "Sample ID: `SRR606249_subset10`"
---

```{r setup, message=FALSE, warning=FALSE}
library(knitr)
library(tidyverse)
library(DT)
opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE)
theme_set(theme_bw())
```


# About {.tabset .tabset-fade}

## About this report

This report pulls in statistics and reports generated after running all workflows as described in the version 1.2 wiki. Each major section (QA, taxonomic classification, assembly, functional annotation) is accessible via a link in the table of contents above. Tabs under each major heading expand to show summary results, links to further reports, details, or documentation on how these results were produced. Within particular tabs there are buttons that further subdivide results that are displayed (e.g., different assemblers, or different AMR tools). 

## About this analysis

Pull some information from here about the samples run, etc. from the logs, command history, etc.


# Read Filtering {.tabset .tabset-fade}

## General statistics

```{r}
read_tsv("SRR606249_subset10_fastqc_multiqc_report_data/multiqc_general_stats.txt") %>% 
  set_names(~str_remove(., "FastQC_mqc-generalstats-")) %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="General QA statistics about samples in this run.")
```


```{r}
get_unique_kmers <- function(file) {
  read_lines(file) %>% 
    grep("unique k-mers", ., value=TRUE) %>% 
    str_extract("\\d+") %>% 
    as.integer()
}
list.files(pattern="uniqueK\\d\\d\\.txt") %>% 
  tibble(file=.) %>% 
  mutate(sample=str_extract(file, ".+trim\\d+")) %>% 
  mutate(K=str_extract(file, "K\\d+")) %>% 
  mutate(number_unique_kmers=map_int(file, get_unique_kmers)) %>% 
  select(-file) %>% 
  mutate_if(is.integer, scales::comma) %>% 
  DT::datatable(caption="Number of unique kmers in each sample in this run.")
```


## MultiQC Reports: FastQC

Click for links to comprehensive quality assessment reports:

```{r, results='asis'}
x <- list.files(pattern="fastqc_multiqc_report\\.html$")
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```



## Detailed FastQC statistics

```{r}
read_tsv("SRR606249_subset10_fastqc_multiqc_report_data/multiqc_fastqc.txt") %>% 
  DT::datatable(caption="Comprehensive information from FastQC (side-scroll to see more)", options = list(scrollX='600px'))
```

```{r}
read_tsv("SRR606249_subset10_fastqc_multiqc_report_data/multiqc_sources.txt") %>% 
  DT::datatable(caption="MultiQC Data Sources")
  
```

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.





# Taxonomic Classification {.tabset .tabset-fade}

## Krona plots {.tabset .tabset-pills}

### Links to existing HTML files

Click for links to krona diagrams:

```{r, results='asis'}
x <- list.files(pattern=".*krona.*html$")
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```

### Direct Krona embedding (experimental)

This is highly experimental. This shows a direct embedding of SRR606249_subset10_trim30.kaiju_species_krona.html.

<iframe src="SRR606249_subset10_trim30.kaiju_species_krona.html" width="1000" height="1000"></iframe>


## Kaiju results {.tabset .tabset-pills}

### Direct links to files

```{r, results='asis'}
# x <- list.files(pattern="kaiju.+out")
# utils:::format.object_size(file.size(x), unit="MB")
# glue::glue("- **[{x}]({x})**") %>% 
#   cat(sep="\n")

```

```{r, results='asis'}
x <- list.files(pattern="kaiju.+summary$")
glue::glue("- **[{x}]({x})**") %>%
  cat(sep="\n")

```

### Inline reports

Coming soon.

## Sourmash Gather Detailed results {.tabset .tabset-pills}

### Trim2 k=21

```{r}
read_csv("SRR606249_subset10_trim2_k21.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=21 (side-scroll to see more)", options = list(scrollX='600px'))
```

### Trim2 k=31

```{r}
read_csv("SRR606249_subset10_trim2_k31.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=31 (side-scroll to see more)", options = list(scrollX='600px'))
```

### Trim2 k=51

```{r}
read_csv("SRR606249_subset10_trim2_k51.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=51 (side-scroll to see more)", options = list(scrollX='600px'))
```

### Trim30 k=21

```{r}
read_csv("SRR606249_subset10_trim30_k21.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=21 (side-scroll to see more)", options = list(scrollX='600px'))
```

### Trim30 k=31

```{r}
read_csv("SRR606249_subset10_trim30_k31.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=31 (side-scroll to see more)", options = list(scrollX='600px'))
```

### Trim30 k=51

```{r}
read_csv("SRR606249_subset10_trim30_k51.gather_output.csv") %>% 
  mutate_if(is.double, round, 2) %>% 
  DT::datatable(caption="Sourmash Gather results for k=51 (side-scroll to see more)", options = list(scrollX='600px'))
```

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.




# Assembly {.tabset .tabset-fade}

## General statistics {.tabset .tabset-pills}

### MEGAHIT

```{r}
read_tsv("SRR606249_subset10.megahit_multiqc_report_data/multiqc_general_stats.txt") %>% 
  set_names(~str_remove(., "QUAST_mqc-generalstats-")) %>% 
  select(Sample, Total_length, N50) %>% 
  DT::datatable(caption="General assembly statistics about samples in this run from MEGAHIT.")
  
```

### MetaSPADES

```{r}
read_tsv("SRR606249_subset10.metaspades_multiqc_report_data/multiqc_general_stats.txt") %>% 
  set_names(~str_remove(., "QUAST_mqc-generalstats-")) %>% 
  select(Sample, Total_length, N50) %>% 
  DT::datatable(caption="General assembly statistics about samples in this run from MetaSPADES.")
```

## QUAST {.tabset .tabset-pills}

### MultiQC reports

Click for links to comprehensive quality assessment reports:

```{r, results='asis'}
x <- list.files(pattern="(megahit|metaspades)_multiqc_report\\.html$")
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```


### Individual QUAST reports

Individual QUAST reports:

```{r, results='asis'}
x <- list.files(pattern="^report\\.html$", recursive=TRUE)
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```


## Detailed results  {.tabset .tabset-pills}

### MEGAHIT

```{r}
read_tsv("SRR606249_subset10.megahit_multiqc_report_data/multiqc_quast.txt") %>% 
  DT::datatable(caption="Comprehensive information from QUAST (side-scroll to see more)", options = list(scrollX='600px'))
  
```

### MetaSPADES

```{r}
read_tsv("SRR606249_subset10.metaspades_multiqc_report_data/multiqc_quast.txt") %>% 
  DT::datatable(caption="Comprehensive information from QUAST (side-scroll to see more)", options = list(scrollX='600px'))
```

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.




# Comparison {.tabset .tabset-fade}

## CSVs

```{r, results='asis'}
x <- list.files(pattern="read_assembly_comparison", recursive=FALSE)
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```

## PNGs

### Direct links to images

```{r, results='asis'}
x <- list.files(pattern="read_assembly_comparison.+\\.png$", recursive=TRUE)
bx <- basename(x)
glue::glue("- **[{bx}]({x})**") %>% 
  cat(sep="\n")
```


### Inline graphics (in the same order as above)

```{r}
knitr::include_graphics(x)
```

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.



# Functional Inference {.tabset .tabset-fade}

## Prokka

Click below to access prokka results:

```{r, results='asis'}
x <- list.files(pattern="podar_metaG_sbu.tsv", recursive=TRUE)
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```

## Antibiotic resistance {.tabset .tabset-pills}

### Abricate (contigs)

```{r, results='asis'}
x <- list.files(pattern="card\\.csv$")
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```

### SRST2 (reads)

```{r, results='asis'}
x <- list.files(pattern="srst2.*results\\.txt$")
glue::glue("- **[{x}]({x})**") %>% 
  cat(sep="\n")
```


## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.






```{r, include=FALSE}
knitr::knit_exit()
```



# xx {.tabset .tabset-fade}

## General statistics

## Detailed results

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.









# xx {.tabset .tabset-fade}

## General statistics

## Detailed results

## Procedure

Information goes here about the procedure, versions, etc. Lorem ipsum dolor sit amet, semper suscipit sea at. Dico scriptorem nec at, ex qui virtute dolores oportere. Duis tantas ponderum ut has. Est saepe mandamus salutatus et, id sed semper detracto moderatius, ei sit aperiam voluptua. Per esse justo fierent eu, duo quando tempor ut. At elitr doming possim vim, ut dolorem appetere nec.



